caffe2/operators/top_k.cu

```diff
@@ -16,216 +16,399 @@
+        CAFFE_CUDA_NUM_THREADS,
+        0,
+        context_.cuda_stream()>>>(
+        indices->template data<TIndex>(),
+        indices->size(),
+        stride,
+        inner_size,
+        k_,
+        flatten_indices->template mutable_data<TIndex>());
   }
-
   return true;
 }
 
 REGISTER_CUDA_OPERATOR(TopK, TopKOp<float, CUDAContext>);
 
-__global__ void fillValuesWithIndicesKernel(
-    const float* values,
-    const TIndex* indices,
-    const TIndex k,
-    const TIndex orignal_last_dim,
-    const TIndex length,
-    float* output) {
-  CUDA_1D_KERNEL_LOOP(i, length) {
-    int first_dim = i / k;
-    int idx = orignal_last_dim * first_dim + indices[i];
-    output[idx] = values[i];
-  }
-}
-
-template <>
-bool TopKGradientOp<float, CUDAContext>::RunOnDevice() {
-  auto& values = Input(0);
-  auto& indices = Input(1);
-  auto& original_input = Input(2);
+template <typename T>
+class TopKGradientOp<T, CUDAContext> : public Operator<CUDAContext> {
+ public:
+  USE_OPERATOR_FUNCTIONS(CUDAContext);
 
-  vector<TIndex> in_dims = values.dims();
+  TopKGradientOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<CUDAContext>(operator_def, ws),
+        OP_SINGLE_ARG(int, "axis", axis_, -1) {}
 
-  // Linearize input tensor except for last dimension
-  // e.g. [3, 4, 5] -> [12, 5]
-  // [5] -> [5]
-  TIndex flatten_shape[] = {size_to_dim_(in_dims.size() - 1, in_dims),
-                            in_dims[in_dims.size() - 1]};
+  ~TopKGradientOp() = default;
 
-  vector<TIndex> original_dims = original_input.dims();
-  auto* output = Output(0);
-  output->Resize(original_dims);
+  bool RunOnDevice() override;
 
-  float* output_data = output->mutable_data<float>();
-  math::Set<float, CUDAContext>(
-      output->size(), float(0), output_data, &context_);
+ private:
+  int axis_;
+};
 
-  int length = flatten_shape[0] * flatten_shape[1];
-  if (length == 0) { // for empty batch
-    return true;
+template <typename T>
+bool TopKGradientOp<T, CUDAContext>::RunOnDevice() {
+  const auto& values = Input(0);
+  const auto& indices = Input(1);
+  const auto& original_input = Input(2);
+  auto* output = Output(0);
+  const std::vector<TIndex>& values_dims = values.dims();
+  const std::vector<TIndex>& origin_dims = original_input.dims();
+  CAFFE_ENFORCE_EQ(values_dims.size(), origin_dims.size());
+  output->Resize(origin_dims);
+  T* output_data = output->template mutable_data<T>();
+  if (axis_ == -1) {
+    axis_ = values_dims.size() - 1;
   }
-
-  int num_threads = std::min(CAFFE_CUDA_NUM_THREADS, length);
-  int blocks = math::divUp(length, num_threads);
-
-  fillValuesWithIndicesKernel<<<
-      blocks,
-      num_threads,
+  const int k = values_dims[axis_];
+  math::Set<T, CUDAContext>(output->size(), T(0), output_data, &context_);
+  const TIndex stride = std::accumulate(
+      values_dims.cbegin() + axis_ + 1,
+      values_dims.cend(),
+      TIndex(1),
+      std::multiplies<TIndex>());
+  SetTopKGradientCUDA<<<
```

Comment:
Nit: briefly add some simple comment about the core algorithm/formula. It will be very helpful.